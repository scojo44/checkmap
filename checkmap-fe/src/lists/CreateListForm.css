.CreateListForm input[name="description"] {
  display: block;
  width: 95%;
}