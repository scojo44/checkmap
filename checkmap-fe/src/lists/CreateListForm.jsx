import React, {useContext} from 'react'
import {useNavigate} from 'react-router-dom';
import {useForm} from 'react-hook-form'
import CheckMapAPI from '../api'
import UserContext from '../UserContext'
import FormField from '../widgets/FormField';
import './CreateListForm.css'

export default function CreateListForm({}) {
  const navigate = useNavigate();
  const {user, setCurrentList, showAlert} = useContext(UserContext);
  const {register, handleSubmit, formState: {errors}} = useForm();

  return (
    <form className="CreateListForm" onSubmit={handleSubmit(async data => await addList(data))}>
      <h2>Create a New List</h2>
      <FormField name="name" label="Name" {...{register, errors}} validation={{required: 'Please enter a name for your list'}}/>
      <FormField name="description" label="Description (optional)" {...{register, errors}}/>
      <FormField inputType="color" name="color" label="Fill Color" {...{register, errors}}/>
      <p>
        <label htmlFor="regionType">Region Type: </label>
        <select {...register("regionType")}>
          <option value="State">States</option>
          <option value="County">Counties</option>
        </select>
        {errors.regionType && <span className="input-error"> {errors.regionType.message}</span>}
      </p>
      <button type="submit">Submit</button>
      <button type="button" onClick={hideForm}>Cancel</button>
    </form>
  );

  async function addList(newList) {
    try {
      const list = await CheckMapAPI.createList(user.username, newList);
      user.lists.push(list);
      setCurrentList(list);
      hideForm();
    }
    catch(e) {
      showAlert('error', 'Error creating list:' + e);
    }
  }

  function hideForm() {
    navigate('/map');
  }
}
