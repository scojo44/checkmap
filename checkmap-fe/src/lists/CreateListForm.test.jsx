import {render, fireEvent, waitFor} from '@testing-library/react'
import {MemoryRouter} from 'react-router-dom'
import CreateListForm from './CreateListForm'

const mockAddList = vi.fn();

describe('CreateListForm Tests', () => {
  beforeEach(() => {
    mockAddList.mockClear();
  });

  it('Renders without crashing', () => {
    render(
      <MemoryRouter>
        <CreateListForm/>
      </MemoryRouter>
    );
  });

  it('Matches snapshot', () => {
    const {asFragment} = render(
      <MemoryRouter>
        <CreateListForm/>
      </MemoryRouter>
    );

    expect(asFragment()).toMatchSnapshot();
  });

  it('Calls the passed function with the submitted values', async () => {
    const {getByText, getByLabelText} = render(
      <MemoryRouter>
        <CreateListForm addNewList={mockAddList}/>
      </MemoryRouter>
    );

    const listName = getByLabelText('Name:');
    const listDesc = getByLabelText('Description (optional):');
    const listType = getByLabelText('Region Type:');
    const listColor = getByLabelText('Fill Color:');
    const button = getByText('Submit');

    fireEvent.change(listName, {target: {value: 'New Test List'}});
    fireEvent.change(listDesc, {target: {value: 'Test description'}});
    fireEvent.change(listType, {target: {value: 'County'}});
    fireEvent.change(listColor, {target: {value: '#9370db'}});
    fireEvent.click(button);

    await waitFor(() => expect(mockAddList).toHaveBeenCalled());
    await waitFor(() => expect(mockAddList).toHaveBeenCalledWith({
      name: 'New Test List',
      description: 'Test description',
      regionType: 'County',
      color: '#9370db'
    }));
  });
});
