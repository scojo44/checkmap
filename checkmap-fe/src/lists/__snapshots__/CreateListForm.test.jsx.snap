// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`CreateListForm Tests > Matches snapshot 1`] = `
<DocumentFragment>
  <form
    class="CreateListForm"
  >
    <h2>
      Create a New List
    </h2>
    <p
      class="FormField"
    >
      <label
        for="name"
      >
        Name: 
      </label>
      <input
        id="name"
        name="name"
      />
    </p>
    <p
      class="FormField"
    >
      <label
        for="description"
      >
        Description (optional): 
      </label>
      <input
        id="description"
        name="description"
      />
    </p>
    <p
      class="FormField"
    >
      <label
        for="color"
      >
        Fill Color: 
      </label>
      <input
        id="color"
        name="color"
        type="color"
      />
    </p>
    <p>
      <label
        for="regionType"
      >
        Region Type: 
      </label>
      <select
        id="regionType"
        name="regionType"
      >
        <option
          value="State"
        >
          States
        </option>
        <option
          value="County"
        >
          Counties
        </option>
      </select>
    </p>
    <button
      type="submit"
    >
      Submit
    </button>
    <button
      type="button"
    >
      Cancel
    </button>
  </form>
</DocumentFragment>
`;
